File: alchemist-implementationbase/src/main/java/it/unibo/alchemist/model/timedistributions/RandomDiracComb.java
/*
 * Copyright (C) 2010-2023, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */

package it.unibo.alchemist.model.timedistributions;

import it.unibo.alchemist.model.Time;
import it.unibo.alchemist.model.times.DoubleTime;
import org.apache.commons.math3.random.RandomGenerator;

import java.io.Serial;

/**
 * A {@link DiracComb} whose rate is determined (uniformly) randomly within the provided bounds.
 *
 * @param <T> concentration type
 */
public class RandomDiracComb<T> extends DiracComb<T> {

    @Serial
    private static final long serialVersionUID = 1L;

    /**
     * @param rng
     *            the {@link RandomGenerator}
     * @param start
     *            the start {@link Time}
     * @param minRate
     *            the minimum rate
     * @param maxRate
     *            the maximum rate
     */
    public RandomDiracComb(final RandomGenerator rng, final Time start, final double minRate, final double maxRate) {
        super(start, minRate + (maxRate - minRate) * rng.nextDouble());
        if (minRate > maxRate || maxRate <= 0 || minRate <= 0) {
            throw new IllegalArgumentException("Invalid rate values: {min=" + minRate + ", max=" + maxRate + "}.");
        }
    }

    /**
     * @param rng
     *            the {@link RandomGenerator}
     * @param minRate
     *            the minimum rate
     * @param maxRate
     *            the maximum rate
     */
    public RandomDiracComb(final RandomGenerator rng, final double minRate, final double maxRate) {
        this(rng, new DoubleTime(minRate * rng.nextDouble()), minRate, maxRate);
        if (minRate > maxRate || maxRate <= 0) {
            throw new IllegalArgumentException("Invalid rate values: {min=" + minRate + ", max=" + maxRate + "}.");
        }
    }
}

File: alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/timedistributions/RandomDiracComb.kt
/*
 * Copyright (C) 2010-2026, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */

package it.unibo.alchemist.model.timedistributions

import it.unibo.alchemist.model.Node
import it.unibo.alchemist.model.Time
import it.unibo.alchemist.model.TimeDistribution
import it.unibo.alchemist.model.times.DoubleTime
import org.apache.commons.math3.random.RandomGenerator

/** A [DiracComb] whose rate is sampled uniformly within the provided bounds for each new instance. */
class RandomDiracComb<T> private constructor(
    private val randomGenerator: RandomGenerator,
    start: Time,
    private val minRate: Double,
    private val maxRate: Double,
    private val randomizeStart: Boolean,
) : DiracComb<T>(start, sampleRate(randomGenerator, minRate, maxRate)) {

    /** Builds a randomized-rate generator with an explicit absolute start time. */
    constructor(randomGenerator: RandomGenerator, start: Time, minRate: Double, maxRate: Double) :
        this(randomGenerator, start, minRate, maxRate, false)

    /** Builds a randomized-rate generator whose initial start is randomized as well. */
    constructor(randomGenerator: RandomGenerator, minRate: Double, maxRate: Double) :
        this(randomGenerator, DoubleTime(minRate * randomGenerator.nextDouble()), minRate, maxRate, true)

    override fun newInstanceOn(node: Node<T>): TimeDistribution<T> = if (randomizeStart) {
        RandomDiracComb(randomGenerator, minRate, maxRate)
    } else {
        RandomDiracComb(randomGenerator, startTime, minRate, maxRate)
    }

    private companion object {
        private fun sampleRate(randomGenerator: RandomGenerator, minRate: Double, maxRate: Double): Double {
            require(minRate <= maxRate && minRate > 0 && maxRate > 0) {
                "Invalid rate values: {min=$minRate, max=$maxRate}."
            }
            return minRate + (maxRate - minRate) * randomGenerator.nextDouble()
        }
    }
}
